src/shared_library.c

```diff
@@ -0,0 +1,204 @@
+  if (lib_symbol == NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting the symbol '%s'. Error '%d'",
+      symbol_name, GetLastError());
+    return NULL;
+  }
+#endif  // _WIN32
+  if (!lib_symbol) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "symbol '%s' does not exist in the library '%s'",
+      symbol_name, lib->library_path);
+    return NULL;
+  }
+  return lib_symbol;
+}
+
+bool
+rcutils_has_symbol(const rcutils_shared_library_t * lib, const char * symbol_name)
+{
+  if (!lib || !lib->lib_pointer || symbol_name == NULL) {
+    return false;
+  }
+
+#ifndef _WIN32
+  // the correct way to test for an error is to call dlerror() to clear any old error conditions,
+  // then call dlsym(), and then call dlerror() again, saving its return value into a variable,
+  // and check whether this saved value is not NULL.
+  dlerror(); /* Clear any existing error */
+  void * lib_symbol = dlsym(lib->lib_pointer, symbol_name);
+  return dlerror() == NULL && lib_symbol != 0;
+#else
+  void * lib_symbol = GetProcAddress(lib->lib_pointer, symbol_name);
+  return GetLastError() == 0 && lib_symbol != 0;
+#endif  // _WIN32
+}
+
+rcutils_ret_t
+rcutils_unload_shared_library(rcutils_shared_library_t * lib)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib->lib_pointer, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib->library_path, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ALLOCATOR(&lib->allocator, return RCUTILS_RET_INVALID_ARGUMENT);
+
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+#ifndef _WIN32
+  // The function dlclose() returns 0 on success, and nonzero on error.
+  int error_code = dlclose(lib->lib_pointer);
+  if (error_code) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("dlclose error: %s", dlerror());
+#else
+  // If the function succeeds, the return value is nonzero.
+  int error_code = FreeLibrary(lib->lib_pointer);
+  if (!error_code) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("FreeLibrary error: %lu", GetLastError());
+#endif  // _WIN32
+    ret = RCUTILS_RET_ERROR;
+  }
+
+  lib->allocator.deallocate(lib->library_path, lib->allocator.state);
+  lib->library_path = NULL;
+  lib->lib_pointer = NULL;
+  lib->allocator = rcutils_get_zero_initialized_allocator();
+  return ret;
+}
+
+rcutils_ret_t
+rcutils_get_platform_library_name(
+  const char * library_name,
+  char * library_name_platform,
+  unsigned int buffer_size)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(library_name, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(library_name_platform, RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (strlen(library_name) > RCUTILS_DIR_PATH_MAX)
+  {
+    RCUTILS_SET_ERROR_MSG(
+      "library_name is too long more than the maximum allowed size\n");
+    return RCUTILS_RET_ERROR;
+  }
+
+  int written = 0;
+
+#ifdef __linux__
+  if (buffer_size > (strlen(library_name) + 7)) {
```

Comment:
Why does the buffer need to be larger and equal size isn't enough?

Reply:
fixed https://github.com/ros2/rcutils/pull/215/commits/5cc468ea931e1ed07a1c4c973bc181904820b8cf